[pytorch] use a mutex in initialize_torch_libraries

rmaz opened the pull request:
Summary: The TORCH_LIBRARY_THREAD_UNSAFE_LAZY_INIT feature is thread unsafe for calling the initializers, but we want to allow the deferred initializer call to be safe from multiple threads. Add a mutex to ensure we have thread safe construction of the libraries post launch.

Differential Revision: D73457714

swolchok commented on aten/src/ATen/core/library.cpp:
```diff
@@ -62,8 +62,10 @@ void Library::reset() {
+  std::mutex torch_library_initializer_mutex;
 } // namespace detail
 void initialize_torch_libraries() {
+  const std::lock_guard<std::mutex> lock(detail::torch_library_initializer_mutex);
```
I was going to suggest doing this via some sort of call_once equivalent (probably initializing a function-local static), but I see why that's not desirable (we might dlopen some code that adds more initializers). Two requests:
1) please document that fact here so nobody comes along and tries to change it
2) it seems inconsistent that we punt thread safety for the write path but not the read path. Why don't we hold this mutex while we write to torch_library_initializers as well?

rmaz replied:
1. will do
2. the reason is that for the write path we expect library loading to be handled by dyld, so already enforced single threaded before application code runs. In this condition holding a lock is unnecessary overhead. The read path can occur on any thread however, and potentially can race, depending on application architecture.

swolchok replied:
do we have any way to guarantee that nobody will screw that up in the future given that we are now relying on it?

rmaz committed:
[pytorch] use a mutex in initialize_torch_libraries (#151938)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/151938

The TORCH_LIBRARY_THREAD_UNSAFE_LAZY_INIT feature is thread unsafe for calling the initializers, but we want to allow the deferred initializer call to be safe from multiple threads. Add a mutex to ensure we have thread safe construction of the libraries post launch.

Differential Revision: D73457714

swolchok reviewed (approved):
OK this is a clear step forward. You have to be careful with the initializer registration path both before and after this PR. Before, initialize wasn't synchronized and now it is.
